File: tests/workflow/producer/rust/src/producer/responses/structf.rs
use super::{identification, producer::Control, protocol, scope::Scope, Context};
use crate::{stat::Alias, stop, test::samples};
use clibri::server;

#[allow(unused_variables)]
pub async fn response<'c, E: server::Error, C: server::Control<E>>(
    request: &protocol::StructF,
    scope: &mut Scope<'_, E, C>,
) -> Result<protocol::StructF, protocol::StructE> {
    let index = scope
        .context
        .get_index(scope.identification.uuid(), Alias::StructF);
    if index == 1 {
        scope
            .context
            .inc_stat(scope.identification.uuid(), Alias::StructF);
        Ok(samples::struct_f::get())
    } else {
        if let Err(err) = scope
            .control
            .events
            .triggerbeaconsemitter(protocol::TriggerBeaconsEmitter {
                uuid: scope.identification.uuid().to_string(),
            })
            .await
        {
            stop!("Fail to emit control.events.structuuid: {}", err);
        }
        scope
            .context
            .inc_stat(scope.identification.uuid(), Alias::StructE);
        Err(samples::struct_e::get())
    }
}

File: tests/workflow/producer/typescript/src/producer/responses/structf.ts
import { Response } from "../implementation/responses/structf";
import {
	Context,
	Producer,
	Identification,
	Filter,
	Protocol,
} from "../implementation/responses";
import { Alias } from "../../stat";
import { panic } from "../../tools";
import { Scope } from "../implementation/scope";

export function response(
	request: Protocol.StructF,
	scope: Scope
): Promise<Response> {
	const stat = scope.context.getStat(scope.consumer.uuid());
	stat.case(Alias.StructF);
	const index = stat.index(Alias.StructF);
	if (index === 0) {
		return Promise.resolve(new Response(Protocol.StructF.defaults()));
	} else if (index === 1) {
		const event = Protocol.TriggerBeaconsEmitter.defaults();
		event.uuid = scope.consumer.uuid();
		scope.producer.events.triggerBeaconsEmitter.emit(event);
		stat.case(Alias.StructE);
		return Promise.resolve(new Response(Protocol.StructE.defaults()));
	} else {
		panic(`Unexpected index for Protocol.StructF request: ${index}`);
		return Promise.reject(
			new Error(`Handler for Protocol.StructF isn't implemented.`)
		);
	}
}
